tests/request_helpers_test.py

```diff
@@ -0,0 +1,408 @@
+        assert ZipkinAttrs(
+            trace_id, span_id, None, '0', True,
+        ) == request_helpers.extract_zipkin_attrs_from_headers({
+            'X-B3-TraceId': trace_id,
+            'X-B3-SpanId': span_id,
+            'X-B3-Sampled': yes_trace_str,
+        }, sample_rate=88.2)
+        assert [] == mock_random.mock_calls
+
+    # non-defer debug-trace, no parent span
+    mock_random.return_value = 0.883  # a "no" if it were used
+    assert ZipkinAttrs(
+        trace_id, span_id, None, '1', True,
+    ) == request_helpers.extract_zipkin_attrs_from_headers({
+        'X-B3-TraceId': trace_id,
+        'X-B3-SpanId': span_id,
+        'X-B3-Flags': '1',
+    }, sample_rate=88.2)
+    assert [] == mock_random.mock_calls
+
+    # non-defer no-trace, yes parent span
+    mock_random.return_value = 0.881  # a "yes" if it were used
+    assert ZipkinAttrs(
+        trace_id, span_id, parent_span_id, '0', False,
+    ) == request_helpers.extract_zipkin_attrs_from_headers({
+        'X-B3-TraceId': trace_id,
+        'X-B3-SpanId': span_id,
+        'X-B3-Sampled': no_trace_str,
+        'X-B3-ParentSpanId': parent_span_id,
+    }, sample_rate=88.2)
+    assert [] == mock_random.mock_calls
+
+    # non-defer yes-trace, yes parent span
+    for yes_trace_str in ('true', '1'):
+        mock_random.return_value = 0.883  # a "no" if it were used
+        assert ZipkinAttrs(
+            trace_id, span_id, parent_span_id, '0', True,
+        ) == request_helpers.extract_zipkin_attrs_from_headers({
+            'X-B3-TraceId': trace_id,
+            'X-B3-SpanId': span_id,
+            'X-B3-Sampled': yes_trace_str,
+            'X-B3-ParentSpanId': parent_span_id,
+        }, sample_rate=88.2)
+        assert [] == mock_random.mock_calls
+
+    # non-defer debug-trace, yes parent span
+    mock_random.return_value = 0.883  # a "no" if it were used
+    assert ZipkinAttrs(
+        trace_id, span_id, parent_span_id, '1', True,
+    ) == request_helpers.extract_zipkin_attrs_from_headers({
+        'X-B3-TraceId': trace_id,
+        'X-B3-SpanId': span_id,
+        'X-B3-Flags': '1',
+        'X-B3-ParentSpanId': parent_span_id,
+    }, sample_rate=88.2)
+    assert [] == mock_random.mock_calls
+
+    # flags (debug) trump Sampled, if they conflict--another edge case
+    mock_random.return_value = 0.883  # a "no" if it were used
+    assert ZipkinAttrs(
+        trace_id, span_id, parent_span_id, '1', True,
+    ) == request_helpers.extract_zipkin_attrs_from_headers({
+        'X-B3-TraceId': trace_id,
+        'X-B3-SpanId': span_id,
+        'X-B3-Flags': '1',
+        'X-B3-Sampled': no_trace_str,
+        'X-B3-ParentSpanId': parent_span_id,
+    }, sample_rate=88.2)
+    assert [] == mock_random.mock_calls
+
+
+@mock.patch('py_zipkin.util.random.random')
+def test_extract_zipkin_attrs_from_headers_single(mock_random):
+    # b3={TraceId}-{SpanId}-{SamplingState}-{ParentSpanId}
+    # where the last two fields are optional.
+
```

Comment:
Do we actually need all these tests to cover all the cases? There's a lot of them and they seem to repeat themselves.

I think the repeating part stems from the fat that we're basically implementing the same logic twice. Could we refactor away the common logic in request_helpers.py so that we only have to test it once? The only difference between the 2 afaict is whether all the fields are in a single header or in multiple ones.

So we could just have the logic to parse the headers t be separate and then call the same underlying function to generate the ZipkinAttrs

Reply:
The specification is a little loose and I kept running into wacky under-specified edge cases.  I think the volume of tests is justified by that.

Personally, I'd rather have overly-exhaustive tests than untested edge cases.

Because of the spec, it's not _quite_ apples to apples.  The single-header form is newer and drops some backwards-compat, the "d" thing is only in single-header, "flags" isn't a thing in single-header.

I guess those impedance mismatches could be smoothed over, but because of the spec, you'd really want the tests like they are to make sure _that_ code is right and each side of the spec (multi-header/single-header) are correct.

Refactoring the implementation to have fewer lines seems reasonable for reasons like clarity, DRY, less bug surface area, etc., but I'm not sure having shorter tests is a great justification.

Let me see what the refactoring looks like/

Reply:
@drolando That was a great idea, and I think the code cleaned up nicely.  Please see latest commit.

I still the the tests are good, as they stand now, for the reasons I mentioned earlier.

Reply:
ps only one flag value was ever defined, 1, so not sure how much under-specifying we'd need to do to clarify d is same as presence of that flag.. can we fix the spec somehow?

Reply:
(I misspoke earlier and said `"flags" isn't a thing in single-header`; I meant `"flags" isn't a thing in multi-header`)

I'm not sure I understand what you're saying.  I think the spec and the code of this PR are clear in that:
1) single-header b3 has a tri-state `SamplingState` segment with "1", "0", and "d" defined where "d" means "debug" and is like a "1" with a bit of additional semantic import.
2) multi-header has an ill-fated `X-B3-Flags` header where `1` means `d` (i.e. debug), and apparently no other flags are defined (not even `0` for "there are no flags" or "debug is not active"?)

As far as fixing the spec, I can only assume other folks have already coded to it and the cat is out of the bag?  Like the true/false thing: I decided to go ahead and be lenient in both encodings here wrt `true`/`false` values of the sampling decision; after the refactoring, either both encodings accept them or neither do.  I decided that being lenient would be better.